### PR TITLE
Cold save image link

### LINKED ISSUE #3: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by grantpaulson6**
Closes #3 

adds image model, #new #create #show controller actions for image

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by grantpaulson6** [comment]
[(onboarding) save link functionality for image sharer](https://trello.com/c/YVWke7wN/988-onboarding-save-link-functionality-for-image-sharer)

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -1,13 +1,59 @@
...
-    get '/'
-    assert_select 'h1', /Welcome*/
+  def test_new
+    get new_image_path
+
+    assert_response :success
+    assert_select '#header', /Create/
+  end
+
+  def test_show
```
Comment (jasonmally) [inline test/controllers/images_controller_test.rb]: You should add a test case for show w/ an invalid id (e.g. get image_url(12343234145))

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -2,4 +2,33 @@ class ImagesController < ApplicationController
...
+      flash[:errors] = @image.errors.messages
+      render :new, status: :unprocessable_entity
+      # QUESTION: would we want to redirect_to new_image_path to maintain consistent URL for client? (shows as
+      # /images after failure)
+    end
+  end
+
+  def show
+    @image = Image.find(params[:id])
+    render
```
Comment (mosessliu) [inline app/controllers/images_controller.rb]: What is `render` doing here?

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,2 +1,7 @@
 module ApplicationHelper
+  def display_errors(field)
+    return [] unless flash[:errors] && flash[:errors][field]
+
+    flash[:errors][field].map { |error| field.capitalize + ' ' + error }
```
Comment (mosessliu) [inline app/helpers/application_helper.rb]: We usually prefer string interpolation
```ruby
{ |error| "#{field.captialize} #{error}" }
```

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -1,13 +1,59 @@
...
+      image_params = { image: { name: '', url: '' } }
+      post images_path, params: image_params
+    end
+
+    # QUESTION: best practice for displaying (wrapper classes with form) and testing each error
+    assert_response :unprocessable_entity
+    assert_select '.name.error', 'Name can\'t be blank'
+    assert_select '.url.error', 'Url must be valid'
   end
+  # test 'test_home' do
```
Comment (jasonmally) [inline test/controllers/images_controller_test.rb]: You're welcome to delete this commented-out test unless you plan on refactoring it for use

File: test/models/image_test.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  def test_image__valid
+    image = Image.new(name: 'zion', url: 'https://www.pewtrusts.org/-/media/post-launch-images/2018/01/istock-479409864/istock-479409864_16x9.jpg?la=en&hash=A70682998CAE7084094117D7CA8E14C340BCEC55')
+
+    # QUESTION: use assert or assert_predicate?
+    assert image.valid?
```
Comment (jasonmally) [inline test/models/image_test.rb]: I've seen both `assert` and `assert_predicate` in the property repo.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -2,4 +2,33 @@ class ImagesController < ApplicationController
   def home
     render
   end
+
+  def new
+    @image = Image.new
+    render
```
Comment (mosessliu) [inline app/controllers/images_controller.rb]: What does calling `render` do here?

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -1,13 +1,59 @@
...
+    assert_equal 'Image url successfully saved.', flash[:notice]
+  end
+
+  def test_create__fail
+    assert_no_difference 'Image.count' do
+      image_params = { image: { name: '', url: '' } }
+      post images_path, params: image_params
+    end
+
+    # QUESTION: best practice for displaying (wrapper classes with form) and testing each error
```
Comment (jasonmally) [inline test/controllers/images_controller_test.rb]: That's a good question! It really seems to depend on how the errors are raised and how they're displayed/stored/

File: app/helpers/application_helper.rb (hunk 3 above)
Comment (grantpaulson6) [reply app/helpers/application_helper.rb]: function is unused, I just deleted it

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -23,7 +20,6 @@ def create
 
   def show
```
Comment (jasonmally) [inline app/controllers/images_controller.rb]: I'd still consider adding a method to validate that 'show' is trying to show with an ID that actually has an associated img

**Pull request merged**